File: WanderIQKit/Sources/WanderIQKit/Sync/Outbox.swift
import Foundation

/// Insertion-ordered, key-coalesced set of pending changes (spec §6.2).
/// One entry per (kind, id); the newest enqueue for a key replaces the older
/// but keeps the original queue position so flush order stays stable.
public struct Outbox: Equatable, Codable, Sendable {
    private var order: [EntityKey] = []
    private var byKey: [EntityKey: PendingChange] = [:]

    public init() {}

    public var pending: [PendingChange] { order.compactMap { byKey[$0] } }

    public mutating func enqueue(_ change: PendingChange) {
        if byKey[change.key] == nil { order.append(change.key) }
        byKey[change.key] = change
    }

    public mutating func acknowledge(_ key: EntityKey) {
        byKey[key] = nil
        order.removeAll { $0 == key }
    }

    public var isEmpty: Bool { byKey.isEmpty }

    // Codable: persist as the ordered pending list; rebuild the index on decode.
    private enum CodingKeys: String, CodingKey { case pending }
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let list = try c.decode([PendingChange].self, forKey: .pending)
        for change in list { enqueue(change) }
    }
    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(pending, forKey: .pending)
    }
}

File: webapp/src/sync/outbox.ts
import { entityKey, type EntityKind, type PendingChange } from "./types";

/** Insertion-ordered, key-coalesced pending changes (protocol §"Outbox"). */
export class Outbox {
  private order: string[] = [];
  private byKey = new Map<string, PendingChange>();

  get pending(): PendingChange[] {
    return this.order.map((k) => this.byKey.get(k)!).filter(Boolean);
  }
  get isEmpty(): boolean {
    return this.byKey.size === 0;
  }

  enqueue(change: PendingChange): void {
    const key = entityKey(change);
    if (!this.byKey.has(key)) this.order.push(key);
    this.byKey.set(key, change);
  }

  acknowledge(ref: { kind: EntityKind; id: string }): void {
    const key = entityKey(ref);
    this.byKey.delete(key);
    this.order = this.order.filter((k) => k !== key);
  }

  toJSON(): PendingChange[] {
    return this.pending;
  }
  static fromJSON(list: PendingChange[]): Outbox {
    const box = new Outbox();
    for (const c of list) box.enqueue(c);
    return box;
  }
}
